### ISSUE TITLE
I18n cop

### ISSUE THREAD

**Issue by tall-dan** [rubocop-hq/rubocop#6141]
**The problem**

I work on a large codebase with many developers. We've started to internationalize our site using [rails i18n](https://guides.rubyonrails.org/i18n.html), but it's difficult to ensure that every feature team is implementing their new features in a compliant way. With a large and somewhat unorganized set of translations, it's also difficult for developers to know if we already have the translations for a given word or phrase.

**A potential solution**
I'd like to create a cop that inspects string assignments. It could be configured to complain at all string assignments (seems noisy unless you've extracted your strings to constants files, which we have) or to complain if there's a string that exactly matches a translated string (looking through `config/locales/**/*.yml`).

The cop would be part of the rails cops, and probably disabled by default. 

I'll need to figure out how to lint `html`, `erb`, `slim`, `haml`, etc. as well as JS (open to suggestions), but that is not a part of this feature suggestion.

**Example**
```ruby
# bad
page_title = 'User Index'

# good
page_title = I18n.t('user_index')
```

**What I'm asking for**
Feedback. Is this a feasible idea? Does it belong as a part of this project? What am I not thinking of?

**Comment by Spone** [comment]
Hi @tall-dan, did you have a look at https://github.com/glebm/i18n-tasks

It's not exactly what you suggest, but it can help in some cases.

**Comment by Edouard-chin** [comment]
We had to internationalize our monolith recently and had to create tools for extracting hardcoded strings from our codebase, one of the tool we used was rucobop.

It was a no go for us to create a cop that checks for hardcoded string assignments because there is too much cases where string assignment were used for business logic stuff and not for displaying something to the user. I believe this isn't specific to our app but to any app, thus I don't think having such a cop is a good idea.

What we ended up doing instead is create a set of cops for methods we know were going to output something to the user (for example Rails' `link_to('hardcodedstring', path))`. We have around 50 custom cops that knows method signature and which arguments are used to add string in the DOM.

Also if you are looking for a tool that lints erb files we created the `erb-lint` gem https://github.com/shopify/erb-lint . It can link html but also embedded ruby, the embedded ruby is passed to rubocop cops so you can reuse any cop you like to lint your erb files.

**Comment by aried3r** [comment]
> What we ended up doing instead is create a set of cops for methods we know were going to output something to the user (for example Rails' link_to('hardcodedstring', path)). We have around 50 custom cops that knows method signature and which arguments are used to add string in the DOM.

Hey, that sounds awesome! Do they happen to be open source as well? [erb-lint](https://github.com/Shopify/erb-lint/blob/master/lib/erb_lint/linters/hard_coded_string.rb) seems to have a cop for views, but I'd love to see those additional cops. 

I'll definitely check out erb-lint, thanks!

**Comment by Edouard-chin** [comment]
> I'll definitely check out erb-lint, thanks!

You are welcome!

The set of cops we created isn't opensourced mainly because most methods that we lint are specific to our internal UI component.

**Comment by aried3r** [comment]
Ah, that's too bad. But I just have to ask, any chance of getting the `link_to` cop to use as a base? ❤️

**Comment by deivid-rodriguez** [comment]
@Edouard-chin What's the status of `erblint`? It has GitHub issues disabled and I don't see any comments from maintainers in the PRs opened by external contributors. So it's maintainance status is a bit of a :question:.

**Comment by EiNSTeiN-** [comment]
I enabled github issues on the repo, I'm unsure why they were disabled to begin with. Probably from when the repo was forked.

**Comment by deivid-rodriguez** [comment]
Wow, that's great! Thanks so much @EiNSTeiN-!

**Comment by Drenmi** [comment]
> Feedback. Is this a feasible idea? Does it belong as a part of this project? What am I not thinking of?

It sounds like a reasonable (and rather useful) tool. However the scope of the cop you're describing is far beyond most existing cops. Perhaps this should be a gem on its own? 🙂

**Comment by tall-dan** [comment]
Thanks to everyone for the feedback! I'm going to stew on this for a bit; I'll post links to anything I create.

**Comment by ZASMan** [comment]
> What we ended up doing instead is create a set of cops for methods we know were going to output something to the user (for example Rails' `link_to('hardcodedstring', path))`. We have around 50 custom cops that knows method signature and which arguments are used to add string in the DOM.

Hi, can you by any chance what you were referring to when you said you had custom cops to check for translations? Like I'm trying to find one that will assure strings in html.erb files are internationalized. Thanks so much!
